src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -600,6 +600,24 @@ MemberDeclarationSyntax adjustStateAndReportStatementOutOfOrder(ref NamespacePar
 
                 return memberOrStatement;
             }
+
+            void parseUsingDirective(ref SyntaxToken openBrace, ref NamespaceBodyBuilder body, ref SyntaxListBuilder initialBadNodes, ref NamespaceParts seen, ref SyntaxListBuilder<MemberDeclarationSyntax> pendingIncompleteMembers)
+            {
+                // incomplete members must be processed before we add any nodes to the body:
+                ReduceIncompleteMembers(ref pendingIncompleteMembers, ref openBrace, ref body, ref initialBadNodes);
+
+                var @using = this.ParseUsingDirective();
+                if (seen > NamespaceParts.Usings)
+                {
+                    @using = this.AddError(@using, ErrorCode.ERR_UsingAfterElements);
```

Comment:
If I recall correctly, adding an order error during parsing (as opposed to binding) has impact on incremental parsing that we generally want to avoid.
On the other hand, it looks like we already do it for top-level statements.
[https://github.com/dotnet/roslyn/blob/master/docs/compilers/Design/Parser.md#do-not-use-ambient-context-to-direct-the-behavior-of-the-parser](https://github.com/dotnet/roslyn/blob/master/docs/compilers/Design/Parser.md#do-not-use-ambient-context-to-direct-the-behavior-of-the-parser)

Reply:
> If I recall correctly, adding an order error during parsing (as opposed to binding) has impact on incremental parsing that we generally want to avoid.

this is true.  however, in this case, this is the right thing to do.  We effectively have no place to put this using-directive as we're now past the location in the CompliationUnitSyntax where we could store them.  So we instead have to break this down and attach to whatever real construct we find.

If we had made CompilationUnitSyntax just have a list of generic member nodes, we could have potentially kept the using directive here and error'ed later.  However, that would have come with the negative of not having a nice strongly typed API for CompilationUnit that explicitly breaks out the different types of children.

Reply:
> If I recall correctly, adding an order error during parsing (as opposed to binding) has impact on incremental parsing that we generally want to avoid.

This is a preexisting behavior of the parser. No change is introduced here. The code is simply extracted into a local function and reused.

---
In reply to: [581664827](https://github.com/dotnet/roslyn/pull/51426#discussion_r581664827) [](ancestors = 581664827)

Reply:
Makes sense. Thanks

---
In reply to: [581718358](https://github.com/dotnet/roslyn/pull/51426#discussion_r581718358) [](ancestors = 581718358)